cs/src/core/ClientSession/AdvancedClientSession.cs

```diff
@@ -737,15 +737,54 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
 
             public bool ConcurrentWriter(ref Key key, ref Value src, ref Value dst, long address)
             {
-                return _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst, address);
+                if (!_clientSession.fht.SupportsMutableIndexes)
```

Comment:
* Make `SupportsMutableIndexes` a field of this struct and see if it helps.
* Other thing to try is to make this struct itself readonly -- will that help.
* Moving else path to non-inlined extra heavy-method might help the fast path?